Panning navigation

leenlab2 opened the pull request:
<!--- Provide a summary of your changes in the Pull Request Title above. -->
<!--- If this is a work in progress (not yet ready to be merged), make this a draft pull request. -->
Replaced the arrow buttons with drag panning navigation for more intuitive control of the graph. 

## Motivation and Context
The previous use of arrow buttons to move the graph around was a bit difficult, and using drag panning has become more of a standard in map-like applications.
<!--- Why is this pull request required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here: -->
<!--- https://docs.github.com/en/github/managing-your-work-on-github/managing-your-work-with-issues-and-pull-requests/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword -->

## Your Changes

<!--- Describe your changes here. -->
**Description**: 

1. Removed arrow buttons and associated methods, and updated the css to push the +/- and reset buttons up
2. Removed the old panning functions
3. Removed `verticalPanFactor` and `horizontalPanFactor` states, added new `viewBox` related states instead
4. Added new panning methods and mouseDown listener to svg
5. Altered methods controlling the attributes of the remainig buttons
6. Documented the methods I added

**Type of change** (select all that apply):
<!--- Put an `x` in all the boxes that apply. -->
<!--- Remove any lines that do not apply. -->

- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Testing

<!--- Please describe in detail how you tested this pull request. -->
<!--- This can include tests you added and manual testing in the web browser. -->
Manual testing in web browser to ensure functionality, and compared with other approaches to find possible bugs.

## Questions and Comments (if applicable)

<!-- Ask any questions you have for the maintainers of this project regarding this PR. -->
<!-- Please describe the steps you have already taken to find the answer to your question. -->
<!-- This will ensure that we can give you clear and relevant advice. -->
<!-- If you have additional comments add them here as well. -->
I wasn't sure what datatype to label the `event` parameter with, so I left it as `{*}`. Also, the zooming does not work in this PR yet, as Kenneth is going to make a PR soon that fixes that. Some of the code I removed, such as the containerWidth part, is going to be in that PR as well, since we faced some related bugs during our initial exploration. As for merge conflicts, it would be best to merge in this PR first as it has most of the base code. 

We decided to keep the +/- buttons as a backup control method since they do that on maps (if someone does not have a working mouse wheel, for example). We agreed among ourselves that removing the arrow keys would be the best UI-wise, but we will still implement `onKeyDown` such that one can control the panning with the keyboard arrow keys instead. That change will also be in Kenneth's PR. However, if you disagree we can undo this change.

As for whether the panning causes issues with selecting nodes, it does not seem to. If you just click, it selects the node, but if you click and drag it pans first then selects the node.

Finally, I did not add any new tests for this since I wanted to wait for your feedback first. 

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have verified that the CircleCI tests have
  passed. <!-- (check after opening pull request) -->
- [ ] I have added tests for my changes. <!-- (delete this checklist item if not applicable) -->

david-yz-liu commented on js/components/graph/Graph.js:
```diff
@@ -41,6 +39,12 @@ export default class Graph extends React.Component {
+      panning: false,
+      panStartX: 0,
+      panStartY: 0,
+      viewBoxPos: {x: 0, y: 0},
```
Let's use one combined dictionary for all 4 viewbox states (x, y, width, height).

david-yz-liu commented on js/components/graph/Graph.js:
```diff
@@ -174,8 +178,7 @@ export default class Graph extends React.Component {
           zoomFactor: 1,
-          horizontalPanFactor: 0,
-          verticalPanFactor: 0,
+          viewBoxPos: {x: 0, y: 0},
```
The previous code sets the viewbox width/height as a max of the container's width/height and the `data.width` and `data.height`, so I think you should do the same thing here.

david-yz-liu commented on js/components/graph/Graph.js:
```diff
@@ -338,6 +341,77 @@ export default class Graph extends React.Component {
+   * @param {*} event
+   */
+  onMouseDownInGraph = event => {
+    document.body.style.cursor = "grab";
```
You can use css for this. Update the `render` method to add a class to one of the container elements when panning, and then use the `cursor` css rule.

david-yz-liu commented on js/components/graph/Graph.js:
```diff
@@ -338,6 +341,77 @@ export default class Graph extends React.Component {
+
+  /**
+   * Initializes the panning process by recording the position of the mouse pointer and adding event listeners.
+   * @param {*} event
```
You can use [`Event`](https://developer.mozilla.org/en-US/docs/Web/API/Event)

david-yz-liu commented on js/components/graph/Graph.js:
```diff
@@ -338,6 +341,77 @@ export default class Graph extends React.Component {
+    document.body.style.cursor = "grab";
+    const {adjustedX, adjustedY} = this.adjustCoordsToViewbox(event.clientX, event.clientY);
+
+    document.getElementById("react-graph").addEventListener("mouseup", this.onMouseUpInGraph);
```
It's not a good practice to add/remove event listeners in React; they should always be attached.

I think you mentioned encountering issues about two different event handlers being called; you can experiment with [`Event.preventDefault`](https://developer.mozilla.org/en-US/docs/Web/API/Event/preventDefault) or perhaps `Event.stopPropagation` to deal with this.

david-yz-liu commented on js/components/graph/Graph.js:
```diff
@@ -338,6 +341,77 @@ export default class Graph extends React.Component {
+   * Pans the graph by moving it in the direction that the mouse moved.
+   * @param {*} event
+   */
+  onMouseMoveInGraph = event => {
```
It is better to use method names that reflect the logical behaviour of the method, not the context you expect it to be called from. So for example, something like `panGraph` is better since it describes the action being taken.

david-yz-liu reviewed (commented):
@leenlab2 very exciting. I left some comments for you to work on. When you're done, please also coordinate with the other Courseography team members so that they can update their branches as well (I recommend just doing a git merge).

leenlab2 commented:
Alright, I made the changes you asked for. Let me know if this is better!
Also, should I start making tests, or should I fix generate.js first? Since generate has been broken since we added `connections` state

david-yz-liu commented:
@leenlab2 you can start writing tests! Why don't you coordinate with the other team members to create a *separate* PR that fixes the generate page (independent of these panning changes).

leenlab2 commented:
Alright, I will! Also, I just implemented the comment about the viewbox width and height, since I missed it in my last commit. We had previously moved this part of the code to one of the zooming methods in Kenneth's PR. I'll get working on the tests now!

leenlab2 commented:
Update: 

- Removed `viewBoxPos` and `viewBoxContainerRatio` and replaced them with `horizontalPanFactor` and `verticalPanFactor`
- Removed the method adjusting coordinates
- Added back in the code that calculates the viewbox attributes in render()

david-yz-liu commented:
Merged as part of #1213.
